background/services/name/resolvers/uns.ts

```diff
@@ -0,0 +1,169 @@
+    }
+  )
+  const data = await response.json()
+
+  return data
+}
+
+/**
+ * Check if a given string a valid UNS domain
+ */
+const IsValidUNSDomainName = (s: string): boolean => {
+  const trimmedString = s.trim()
+  let isValidUNSDomain = false
+
+  // Any valid domain name will have a period as part of the string
+  if (trimmedString.lastIndexOf(".") > 0) {
+    const domainExtension = trimmedString.slice(trimmedString.lastIndexOf("."))
+    const supportedUNSDomains = [
+      ".crypto",
+      ".coin",
+      ".wallet",
+      ".bitcoin",
+      ".888",
+      ".nft",
+      ".dao",
+      ".zil",
+      ".x",
+    ]
+
+    if (supportedUNSDomains.includes(domainExtension)) {
+      isValidUNSDomain = true
+    }
+  }
+  return isValidUNSDomain
+}
+
+export default function unsResolver(): NameResolver<"UNS"> {
+  return {
+    type: "UNS",
+    canAttemptNameResolution(): boolean {
+      return true
+    },
+    canAttemptAvatarResolution(): boolean {
+      return true
+    },
+    canAttemptAddressResolution({ name, network }: NameOnNetwork): boolean {
+      return sameNetwork(network, ETHEREUM) && IsValidUNSDomainName(name)
+    },
+
+    async lookUpAddressForName({
+      name,
+      network,
+    }: NameOnNetwork): Promise<AddressOnNetwork | undefined> {
+      // We try to resolve the name using unstoppable domains resolution
+      const data = await lookupUNSDomain(name)
+      const address = data.meta.owner
+
+      if (address === undefined || address === null) {
+        return undefined
+      }
+
+      return {
+        address,
+        network,
+      }
+    },
+    async lookUpAvatar(
+      addressOrNameOnNetwork: AddressOnNetwork | NameOnNetwork
+    ) {
+      const { network } = addressOrNameOnNetwork
+      const { address } =
+        "address" in addressOrNameOnNetwork
+          ? addressOrNameOnNetwork
+          : { address: undefined }
+
+      if (address === undefined) {
+        return undefined
+      }
+
+      // We initialize the avatarUrn
+      let avatarUrn = ""
+
+      // Then we get all the records associated with the particular ETH address
+      const data = await reverseLookupAddress(address)
+
+      // check if any of the records returned has an NFT picture associated with it
+      const unsRecords = data.data
+
+      if (unsRecords.length > 0) {
+        unsRecords.forEach(
+          (record: { attributes: { records: { [x: string]: never } } }) => {
+            const nftPicture = record.attributes.records["social.picture.value"]
+            if (nftPicture !== undefined || nftPicture !== null) {
+              avatarUrn = nftPicture
+            }
+          }
+        )
+      }
+
+      if (avatarUrn === undefined || avatarUrn === null) {
+        return undefined
+      }
+      // modify the nft picture value and make it a correct avatar urn
+      avatarUrn = avatarUrn.replace("1/erc721", "eip155:1/erc721")
+      avatarUrn = avatarUrn.replace("1/erc1155", "eip155:1/erc721")
```

Comment:
Believe we should anchor these to the start so we don't replace a valid URN should one appear, plus I think we can use a capturing group to do this once:

```suggestion
      avatarUrn = avatarUrn.replace(/^1\/(erc721| erc1155)/, "eip155:1/$1")
```

Wondering also if we should allow NameService to accept erc1155 strings so we don't have to impersonate erc721 here...

Anyway, only the starting anchor is blocking here I think.